### PR TITLE
Reconcile feature flag extensions

### PR DESCRIPTION
**Pull request by vojtechszocs**
This PR aligns `ModelFeatureFlag` and `ActionFeatureFlag` extensions and their usage.

By design, every instance of ^^ represents a plugin's intent to _add new_ Console feature flag.

To validate this invariant, the corresponding extension test (`features.spec.ts`) is updated to check for duplicate flag declarations.

@alecmerdler @jeff-phillips-18 I've made two major changes in the OLM plugin:

1. Removing base flags from OLM `Flags` enum
```diff
 export enum Flags {
   OPERATOR_LIFECYCLE_MANAGER = 'OPERATOR_LIFECYCLE_MANAGER',
-  CAN_LIST_PACKAGE_MANIFEST = 'CAN_LIST_PACKAGE_MANIFEST',
-  CAN_LIST_OPERATOR_GROUP = 'CAN_LIST_OPERATOR_GROUP',
 }
```
since they are already declared in Console `FLAGS` enum.

2. Removing `CAN_LIST_PACKAGE_MANIFEST` flag extension
```diff
-  {
-    type: 'FeatureFlag/Model',
-    properties: {
-      model: models.PackageManifestModel,
-      flag: Flags.CAN_LIST_PACKAGE_MANIFEST,
-    },
-  },
```
since it basically overrides the base SSAR check in `public/actions/features.ts`
```ts
{
  // TODO: Move into OLM plugin
  flag: FLAGS.CAN_LIST_PACKAGE_MANIFEST,
  resourceAttributes: {
    group: 'operators.coreos.com',
    resource: 'packagemanifests',
    verb: 'list',
  },
},
```

@bipuladh @rawagner `ExtensionRegistry` is modified to support the `disallowed` property. Dashboard related extensions should now support both `disallowed` and `required`.

Unknown flag check in `public/reducers/features.ts` is updated to take plugin contributed flags into account.

cc @spadgett @christianvogt

### CODE REVIEW

**Comment by vojtechszocs** [comment]
@spadgett Adding SelfSubjectAccessReview feature flag extension would be the next follow-up step, which would allow us to address `TODO: Move into OLM plugin` comments in base `ssarChecks`.

**Comment by spadgett** [comment]
/hold
for #3747

File: frontend/packages/console-plugin-sdk/src/registry.ts [hunk 1]
```diff
@@ -43,34 +42,35 @@ export class ExtensionRegistry {
...
-      .filter((e) => e.properties.required)
+      .filter((e) => e.properties.required || e.properties.disallowed)
       .reduce(
-        (requiredFlags, e) => _.uniq([...requiredFlags, ..._.castArray(e.properties.required)]),
+        (gatingFlags, e) =>
+          _.uniq([
+            ...gatingFlags,
+            ..._.castArray(e.properties.required),
+            ..._.castArray(e.properties.disallowed),
+          ]),
```
Comment (bipuladh) [inline frontend/packages/console-plugin-sdk/src/registry.ts]:
```
.reduce( (gatingFlags, ext) => ext.properties.required || ext.properties.disallowed ? 
    _.uniq(....)
: gatingFlags)
```
maybe only use reduce?

File: frontend/packages/console-plugin-sdk/src/typings/features.ts [hunk 2]
```diff
@@ -3,14 +3,17 @@ import { K8sKind } from '@console/internal/module/k8s';
...
 
-  export interface ActionFeatureFlag {
+  export interface ModelFeatureFlag extends FeatureFlag {
+    /** If a CRD for this model exists, the feature will be enabled. */
+    model: K8sKind;
+  }
+
+  export interface ActionFeatureFlag extends FeatureFlag {
     /** Function used to detect the feature and set flag name/value via Redux action dispatch. */
     detect: (dispatch: Dispatch) => Promise<any>;
```
Comment (bipuladh) [inline frontend/packages/console-plugin-sdk/src/typings/features.ts]: Some plugins might need to support k8sWatch which does not return a promise.
```suggestion
    detect: (dispatch: Dispatch) => Promise<any> | void;
```

File: frontend/packages/console-plugin-sdk/src/registry.ts (hunk 1 above)
Comment (vojtechszocs) [reply frontend/packages/console-plugin-sdk/src/registry.ts]: Sure, I can put the `filter` logic inside the `reduce` handler.

**Comment by vojtechszocs** [comment]
@bipuladh @cloudbehl Regarding Bipul's https://github.com/openshift/console/pull/3987#discussion_r367955125 - I've discussed it with @rawagner, using `k8sWatch` with this particular extension type can lead to problems.

On a vanilla k8s cluster or a cluster without OCS capability, CRD for `OCSServiceModel` is missing which causes `k8sWatch` for `OCSServiceModel` CR instance to be repeatedly failing (due to missing CRD) and causing error messages over time.

Detection of various platforms (OpenShift, Baremetal, etc.) typically uses `coFetchJSON` or `k8sGet` which are basically one-time detections. For them, using the `ActionFeatureFlag` extension is a good fit.

However, in case of OCS "independent" flag, driven by a specific `OCSServiceModel` CR instance:
```ts
k8sWatch(OCSServiceModel, { ns: 'openshift-storage', name: 'ocs-storagecluster' })
```

I can see two solutions here:

1. One-time detection of OCS "independent" flag (not via websocket) - if we expect the specific `OCSServiceModel` CR instance to be always present on an OCS-enabled cluster.

2. Use `ModelFeatureFlag` extension to detect presence of `OCSServiceModel` CRD, then use a (to be implemented) new extension type whose detector function executes whenever the specified Console flags change. In other words, first detect the presence of a CRD and then, once we know the CRD exists on the cluster, open a websocket and use it to drive the actual OCS "independent" flag.

**Comment by vojtechszocs** [comment]
Rebased and addressed https://github.com/openshift/console/pull/3987#discussion_r367953104

**Comment by spadgett** [comment]
/hold
the merge queue is blocked

**Pull request merged**